cincinnati/src/plugins/mod.rs

```diff
@@ -0,0 +1,285 @@
+        PluginIO::InternalIO(internal_io)
+    }
+}
+
+/// Convert from ExternalIO to PluginIO
+///
+/// This merely wraps the struct into the enum variant.
+impl From<ExternalIO> for PluginIO {
+    fn from(external_io: ExternalIO) -> Self {
+        PluginIO::ExternalIO(external_io)
+    }
+}
+
+/// Try to convert from ExternalIO to InternalIO
+///
+/// This can fail because the ExternalIO bytes need to be deserialized into the
+/// InternalIO type.
+impl TryFrom<ExternalIO> for InternalIO {
+    type Err = Error;
+
+    fn try_from(external_io: ExternalIO) -> Fallible<Self> {
+        let mut plugin_exchange: interface::PluginExchange =
+            protobuf::parse_from_bytes(&external_io.bytes)
+                .context("could not parse ExternalIO to PluginExchange")?;
+
+        Ok(Self {
+            graph: plugin_exchange.take_graph().into(),
+            parameters: Default::default(),
+        })
+    }
+}
+
+/// Convert from an IntenralIO to a PluginExchange
+///
+/// This conversion cannot fail.
+impl From<InternalIO> for interface::PluginExchange {
+    fn from(internal_io: InternalIO) -> Self {
+        let mut plugin_exchange = Self::new();
+
+        plugin_exchange.set_graph(internal_io.graph.into());
+        plugin_exchange.set_parameters(internal_io.parameters);
+
+        plugin_exchange
+    }
+}
+
+/// Convert from InternalIO to ExternalIO
+///
+/// The serialization can apparently fail.
+impl TryFrom<InternalIO> for ExternalIO {
+    type Err = Error;
+
+    fn try_from(internal_io: InternalIO) -> Fallible<Self> {
+        use protobuf::Message;
+        let plugin_exchange: interface::PluginExchange = internal_io.into();
+        let bytes: Vec<u8> = plugin_exchange.write_to_bytes()?;
+        Ok(Self { bytes })
+    }
+}
+
+/// Try to convert from PluginIO to InternalIO
+///
+/// In case of the variant Plugion::ExternalIO this involves deserialization
+/// which might fail.
+impl TryFrom<PluginIO> for InternalIO {
+    type Err = Error;
+
+    fn try_from(plugin_io: PluginIO) -> Fallible<Self> {
+        match plugin_io {
+            PluginIO::InternalIO(internal_io) => Ok(internal_io),
+            PluginIO::ExternalIO(external_io) => external_io.try_into(),
+        }
+    }
+}
+
+/// Copmatibility impl to automatically convert the unvarianted PluginIO when the
+/// variant struct ExternalIO is expented.
+///
+/// This may fail due to the possible failing conversion.
+impl TryFrom<PluginIO> for ExternalIO {
+    type Err = Error;
+
+    fn try_from(plugin_io: PluginIO) -> Fallible<Self> {
+        let final_io = match plugin_io {
+            PluginIO::ExternalIO(external_io) => external_io,
+            PluginIO::InternalIO(internal_io) => internal_io.try_into()?,
+        };
+        Ok(final_io)
+    }
+}
+
+/// Wrapper struct for the implementation for implementing Plugin<PluginIO>
+pub struct InternalPluginWrapper<T>(pub T);
+
+/// Wrapper struct for the implementation for implementing Plugin<PluginIO>
```

Comment:
Here and in struct above, is `for the implementation` a leftover?

Reply:
It supposedly justifies the existence of these structs, as they're only used for that. Did it confuse you?

Reply:
> It supposedly justifies the existence of these structs, as they're only used for that. Did it confuse you?

Please ignore me ;-)